homeassistant/components/overkiz/update.py

```diff
@@ -0,0 +1,113 @@
+"""Support for Overkiz updates."""
+from __future__ import annotations
+
+from typing import Any
+
+from pyoverkiz.enums import UIClass, UIWidget, UpdateBoxStatus
+from pyoverkiz.models import Gateway
+
+from homeassistant.components.update import (
+    UpdateDeviceClass,
+    UpdateEntity,
+    UpdateEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import HomeAssistantOverkizData
+from .const import DOMAIN
+from .entity import OverkizEntity
+
+UPDATE_COMMAND = "update"
+
+UPDATE_AVAILABLE_STATUS = {
+    UpdateBoxStatus.READY_TO_UPDATE,
+    UpdateBoxStatus.READY_TO_BE_UPDATED_BY_SERVER,
+    UpdateBoxStatus.READY_TO_UPDATE_LOCALLY,
+    UpdateBoxStatus.READY_TO_UPDATE_LOCALLY,
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the Overkiz update from a config entry."""
+    data: HomeAssistantOverkizData = hass.data[DOMAIN][entry.entry_id]
+
+    entities: list[OverkizUpdateEntity] = []
+
+    for device in data.platforms[Platform.UPDATE]:
+        if device.widget == UIWidget.POD or device.ui_class == UIClass.POD:
+            for command in device.definition.commands:
+                if command.command_name == UPDATE_COMMAND:
+                    entities.append(
+                        OverkizUpdateEntity(
+                            device.device_url,
+                            data.coordinator,
+                        )
+                    )
+
+    async_add_entities(entities)
+
+
+class OverkizUpdateEntity(OverkizEntity, UpdateEntity):
+    """Define a Overkiz update entity."""
+
+    _attr_device_class = UpdateDeviceClass.FIRMWARE
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL | UpdateEntityFeature.PROGRESS
+    )
+
+    def get_gateway(self) -> Gateway | None:
+        """Get associated gateway."""
+
+        for gateway in self.coordinator.client.gateways:
+            if gateway.id == self.device.gateway_id:
+                return gateway
+
+        return None
+
+    @property
+    def installed_version(self) -> str | None:
+        """Return version currently installed."""
+        gateway = self.get_gateway()
+
+        if gateway:
+            return gateway.connectivity.protocol_version
+
+        return None
+
+    @property
+    def latest_version(self) -> str | None:
+        """Return latest available version."""
+        gateway = self.get_gateway()
+
+        if (
+            gateway
+            and gateway.update_status in UPDATE_AVAILABLE_STATUS
+            and self.installed_version is not None
+        ):
+            return "new version available"
```

Comment:
<img width="241" alt="image" src="https://user-images.githubusercontent.com/1424596/206034098-b65127e1-5dd2-4683-8e3e-1dd259080a62.png">

Unfortunately Somfy does not provide the new version. Can we leave this empty? How do other integrations do this?

Reply:
The update entity can't be used in that case at this point.

Reply:
@frenck how could we make it work in HA world?  
I'll marke this PR as draft in the mean time.

Reply:
As I said, you can't. It requires the latest version.

Reply:
@nyroDev I see you keep rebasing this PR and bringing it up to date with latest core standards. If you are still interested in this feature, you could see if other integrations would benefit of such a change and make a proposal at https://github.com/home-assistant/architecture/discussions.

Reply:
@frenck I can see that you made the same comment [on another PR](https://github.com/home-assistant/core/pull/75391/files#r973683525) that came [to production](https://github.com/home-assistant/core/blob/dev/homeassistant/components/advantage_air/update.py#L53).
After reading the code of update, I can see 2 problems why using a string is a bad idea:
1. [_version_is_newer](https://github.com/home-assistant/core/blob/dev/homeassistant/components/update/__init__.py#L187) will probably not work correctly
2. [Skipping version](https://github.com/home-assistant/core/blob/dev/homeassistant/components/update/__init__.py#L312) will basically skip any new updates

Did I miss another point?

One hack that I can do on this PR might be to return a version build from the current Date "2023.11.0" to match the corresponding format of overkiz version.  
That will works correctly with 1 downside: if the user skip it, it will come back the first day of the next month.

Otherwise, at a higher level, maybe we can :  
- Allow latest_version to be set as a new constant `NEW_VERSION_AVAILABLE_NOT_KNOWN`
- If that's the case, the _version_is_newer should be skipped, considering always new
- If that's the case, skipping it could store as skipped version "VERSION_NOT_KNOW_YYYY.MM" where YYYY.MM corresponding to the current month.
- The skipped version would have been rewritten to handle this specific case.

What do you think?